python/lsst/rubintv/analysis/service/database.py

```diff
@@ -435,50 +456,75 @@ def query(
         columns: list[str],
         query: Query | None = None,
         data_ids: list[tuple[int, int]] | None = None,
-    ) -> dict[str, list]:
-        """Query a table and return the results
+        aggregator: str | None = None,
+    ) -> dict[str, list] | dict[str, int]:
+        """
+        Query a table and return the results.
 
         Parameters
         ----------
-        columns :
+        columns : list[str]
             The ``table.column`` names of the columns to load.
-        query :
-            A query used on the table.
-            If `query` is ``None`` then all the rows
-            in the query are returned.
-        data_ids :
+        query : Query | None
+            A query used on the table. If `query` is None, all rows are
+            returned.
+        data_ids : list[tuple[int, int]] | None
             The data IDs to query, in the format ``(day_obs, seq_num)``.
+        aggregator : str | None
+            The SQL aggregation function to apply (e.g., 'sum', 'avg').
 
         Returns
         -------
-        result :
-            A dictionary of columns as keys and lists of values as values.
+        result : dict[str, list] | dict[str, int]
+            A dictionary of columns as keys and lists of values or aggregated
+            results as values.
         """
         # Get the models for the columns
-        table_columns, table_names, data_id_columns = self.get_column_models(columns)
-        day_obs_column, seq_num_column = data_id_columns
+        table_columns, table_names, data_id_columns = self.get_column_models(columns, aggregator is not None)
+        if data_id_columns:
+            day_obs_column, seq_num_column = data_id_columns
 
-        logger.info(f"table names: {table_names}")
+        logger.info(f"Table names: {table_names}")
 
-        # generate the query
+        # Generate the base query
         query_model = sqlalchemy.and_(*[col.isnot(None) for col in table_columns])
+        select_from = self.joins.build_join(table_names)
```

Comment:
I don't know if this select clause can be here. It's possible that a query might use columns that were not in the original list of columns, so I'm not sure if the joins can be properly built in that case (for example imagine the query was over ra/dec but the column being searched is something else. That's why if query is not None we have to add the table names before querying.

Reply:
I can see how this might be a problem for the aggregator. Maybe cache the table names before the query and data_id table names are optionally added and use that in your select clause and return dict?

Reply:
I've moved the table joining to below the custom query logic so that all tables and columns are now included in the FROM clause.